src/persist/src/indexed/background.rs

```diff
@@ -154,68 +155,160 @@ impl<B: Blob> Maintainer<B> {
-            batches.push(
-                blob.get_trace_batch_async(key, CacheHint::NeverAdd)
-                    .recv()?,
-            );
-        }
+        // Holding pen for data before it gets consolidated and placed into
+        // a ColumnarRecords
+        let mut updates = vec![];
+        let mut builder = ColumnarRecordsVecBuilder::default();
 
-        for batch in batches.iter() {
-            updates.extend(batch.updates.iter().flat_map(|u| u.iter()));
-        }
+        let mut new_size_bytes = 0;
+        let mut keys = vec![];
+        let format = ProtoBatchFormat::ParquetKvtd;
 
-        for ((_, _), ts, _) in updates.iter_mut() {
-            ts.advance_by(desc.since().borrow());
+        loop {
+            // Pull more data if either of the holding pens from first or second are empty
+            if first_updates.is_empty() {
+                if let Some(first) = first_batch_iter.next() {
+                    let first_batch = blob
+                        .get_trace_batch_async(first, CacheHint::NeverAdd)
+                        .recv()?;
+                    first_updates.extend(first_batch.updates.iter().flat_map(|u| {
+                        u.iter()
+                            .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+                    }));
+                }
+            }
+
+            if second_updates.is_empty() {
+                if let Some(second) = second_batch_iter.next() {
+                    let second_batch = blob
+                        .get_trace_batch_async(second, CacheHint::NeverAdd)
+                        .recv()?;
+                    second_updates.extend(second_batch.updates.iter().flat_map(|u| {
+                        u.iter()
+                            .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+                    }));
+                }
+            }
+
+            // Figure out the stopping point for records we can safely consolidate.
+            // This is the min max value across the two batches we have seen so far.
+            let threshold_key_val = match (first_updates.last(), second_updates.last()) {
+                (Some((f, _, _)), Some((s, _, _))) => {
+                    if f < s {
+                        f.clone()
+                    } else {
+                        s.clone()
+                    }
+                }
+                (Some((f, _, _)), None) => f.clone(),
+                (None, Some((s, _, _))) => s.clone(),
+                (None, None) => break,
+            };
+
+            // Drain all of the records <= threshold out of the various holding
+            // pens.
+            //
+            // TODO: this could be replaced with a drain_filter if that wasn't
+            // experimental.
+            let mut i = 0;
+            while i < first_updates.len() {
+                if first_updates[i].0 <= threshold_key_val {
+                    let data = first_updates.remove(i);
```

Comment:
doesn't this do O(n^2) memcpys?

my immediate reaction is maybe VecDeque helps but I haven't thought about it enough to be sure

Reply:
this is a hell of a catch thank you! I think `split_at` does what we want more efficiently I'll test out all options and add a new microbenchmark

Reply:
Confirmed - this had terrible performance. The 50MiB compact trace benchmark which takes ~150ms per iteration with `split_at` took ~75 seconds per iteration with the "copy a quadratic number of bytes by repeatedly calling remove" implementation. I didn't bother to analyze that more closely as the single benchmark run took over 12 minutes.